xds/src/main/java/io/grpc/xds/sds/SdsClient.java

```diff
@@ -0,0 +1,308 @@
+  }
+
+  /**
+   * Gets the singleton instance after constructing it as needed.
+   *
+   * @param configSource the {@link ConfigSource} pointing to the SDS server.
+   * @param node Node id of this client.
+   */
+  static synchronized SdsClient getInstance(ConfigSource configSource, Node node) {
+    checkNotNull(configSource, "configSource");
+    checkNotNull(node, "node");
+    if (instance == null) {
+      instance = new SdsClient(configSource, node);
+      return instance;
+    }
+    // check if apiConfigSource match
+    if (instance.configSource.equals(configSource) && instance.clientNode.equals(node)) {
+      return instance;
+    }
+    throw new UnsupportedOperationException(
+        "Multiple SdsClient's with different ApiConfigSource and Node's not supported");
+  }
+
+  private SdsClient(ConfigSource configSource, Node node) {
+    checkNotNull(configSource, "configSource");
+    checkNotNull(node, "node");
+    this.configSource = configSource;
+    this.clientNode = node;
+    this.targetUri = extractUdsTarget(configSource);
+  }
+
+  /** Create the client with given configSource, node and channel. */
+  @VisibleForTesting
+  SdsClient(ConfigSource configSource, Node node, ManagedChannel channel) {
+    this(configSource, node);
+    this.channel = channel;
+  }
+
+  @VisibleForTesting
+  static String extractUdsTarget(ConfigSource configSource) {
+    checkArgument(
+        configSource.hasApiConfigSource(), "only configSource with ApiConfigSource supported");
+    ApiConfigSource apiConfigSource = configSource.getApiConfigSource();
+    checkArgument(
+        ApiType.GRPC.equals(apiConfigSource.getApiType()),
+        "only GRPC ApiConfigSource type supported");
+    checkArgument(
+        apiConfigSource.getGrpcServicesCount() == 1,
+        "expecting exactly 1 GrpcService in ApiConfigSource");
+    GrpcService grpcService = apiConfigSource.getGrpcServices(0);
+    checkArgument(
+        grpcService.hasGoogleGrpc() && !grpcService.hasEnvoyGrpc(),
+        "only GoogleGrpc expected in GrpcService");
+    GoogleGrpc googleGrpc = grpcService.getGoogleGrpc();
+    // for now don't support any credentials
+    checkArgument(
+        !googleGrpc.hasChannelCredentials()
+            && googleGrpc.getCallCredentialsCount() == 0
+            && Strings.isNullOrEmpty(googleGrpc.getCredentialsFactoryName()),
+        "No credentials supported in GoogleGrpc");
+    String targetUri = googleGrpc.getTargetUri();
+    checkArgument(!Strings.isNullOrEmpty(targetUri), "targetUri in GoogleGrpc is empty!");
+    return targetUri;
+  }
+
+  /** Response observer for SdsClient. */
+  private final class ResponseObserver implements StreamObserver<DiscoveryResponse> {
+
+    ResponseObserver() {}
+
+    @Override
+    public void onNext(DiscoveryResponse discoveryResponse) {
+      try {
+        processDiscoveryResponse(discoveryResponse);
+      } catch (Exception e) {
+        logger.log(Level.SEVERE, "processDiscoveryResponse", e);
```

Comment:
Actually this should be handled with a [NACK](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#ack-nack-and-versioning) as per  and [error-detail](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/discovery.proto#envoy-api-field-discoveryrequest-error-detail) instead of calling `onError` (which will terminate the stream). Will make that change (after multiple client change)